core/drivers/crypto/caam/acipher/caam_ecc.c

```diff
@@ -0,0 +1,724 @@
+
+	/* Get the number of bytes of d to pad with 0's */
+	d_size = crypto_bignum_num_bytes(inkey->d);
+	crypto_bignum_bn2bin(inkey->d, outkey->d.data + size_sec - d_size);
+
+	cache_operation(TEE_CACHECLEAN, outkey->d.data, outkey->d.length);
+
+	return CAAM_NO_ERROR;
+}
+
+/*
+ * Convert TEE ECC Curve to CAAM ECC Curve
+ *
+ * @tee_curve  TEE ECC Curve
+ */
+static enum caam_ecc_curve get_caam_curve(uint32_t tee_curve)
+{
+	enum caam_ecc_curve caam_curve = CAAM_ECC_UNKNOWN;
+
+	if (tee_curve > 0 &&
+	    tee_curve < CAAM_ECC_MAX + TEE_ECC_CURVE_NIST_P192) {
+		/*
+		 * Realign TEE Curve assuming NIST_P192 is the first entry in
+		 * the list of supported ECC curves.
+		 */
+		caam_curve = tee_curve - TEE_ECC_CURVE_NIST_P192
+			     + CAAM_ECC_P192;
+	}
+
+	return caam_curve;
+}
+
+/*
+ * Allocate a ECC keypair
+ *
+ * @key        Keypair
+ * @size_bits  Key size in bits
+ */
+static TEE_Result do_allocate_keypair(struct ecc_keypair *key, size_t size_bits)
+{
+	ECC_TRACE("Allocate Keypair of %zu bits", size_bits);
+
+	/* Initialize the key fields to NULL */
+	memset(key, 0, sizeof(*key));
+
+	/* Allocate Secure Scalar */
+	key->d = crypto_bignum_allocate(size_bits);
+	if (!key->d)
+		goto out;
+
+	/* Allocate Public coordinate X */
+	key->x = crypto_bignum_allocate(size_bits);
+	if (!key->x)
+		goto out;
+
+	/* Allocate Public coordinate Y */
+	key->y = crypto_bignum_allocate(size_bits);
+	if (!key->y)
+		goto out;
+
+	return TEE_SUCCESS;
+
+out:
+	ECC_TRACE("Allocation error");
+
+	crypto_bignum_free(key->d);
+	crypto_bignum_free(key->x);
+
+	return TEE_ERROR_OUT_OF_MEMORY;
+}
+
+/*
+ * Allocate an ECC Public Key
+ *
+ * @key        Public Key
+ * @size_bits  Key size in bits
+ */
+static TEE_Result do_allocate_publickey(struct ecc_public_key *key,
+					size_t size_bits)
+{
+	ECC_TRACE("Allocate Public Key of %zu bits", size_bits);
+
+	/* Initialize the key fields to NULL */
+	memset(key, 0, sizeof(*key));
+
+	/* Allocate Public coordinate X */
+	key->x = crypto_bignum_allocate(size_bits);
+	if (!key->x)
+		goto out;
+
+	/* Allocate Public coordinate Y */
+	key->y = crypto_bignum_allocate(size_bits);
+	if (!key->y)
+		goto out;
+
+	return TEE_SUCCESS;
+
+out:
+	ECC_TRACE("Allocation error");
+
+	crypto_bignum_free(key->x);
+
+	return TEE_ERROR_OUT_OF_MEMORY;
+}
+
+/*
+ * Free an ECC public key
+ *
+ * @key  Public Key
+ */
+static void do_free_publickey(struct ecc_public_key *key)
+{
+	crypto_bignum_free(key->x);
+	crypto_bignum_free(key->y);
+}
+
+/*
+ * Generate ECC keypair
+ *
+ * @key        [out] Keypair
+ * @key_size   Key size in bits multiple of 8 bits
+ */
+static TEE_Result do_gen_keypair(struct ecc_keypair *key, size_t key_size)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	enum caam_status retstatus = CAAM_FAILURE;
+	enum caam_ecc_curve curve = CAAM_ECC_UNKNOWN;
+	struct caambuf d = { };
+	struct caambuf xy = { };
+	struct caam_jobctx jobctx = { };
+	uint32_t *desc = NULL;
+	uint32_t desclen = 0;
+
+#ifdef CFG_CAAM_64BIT
```

Comment:
I'd prefer to move this logic closer to the top of this file same for the other cases where they also could be combined.

Reply:
Ok